earth_enterprise/src/common/geFilePoolImpl.h

```diff
@@ -20,18 +20,49 @@
 
 #include <khMTTypes.h>
 #include <fcntl.h>
+#include <string>
+#include <memory>
 
 class geFilePool;
 
+class AbstractFileAccessor {
+public:
+  virtual bool isValid() { return false; }
+  virtual void invalidate() {};
+  virtual void setFD(int i) {};
+  virtual int getAsFD() { return -1; }
+
+  virtual int Open(const std::string &fname, int flags, mode_t createMask) = 0;
+  virtual int FsyncAndClose() = 0;
+  virtual int Close() = 0;
+  virtual bool PreadAll(void* buffer, size_t size, off64_t offset) = 0;
+  virtual bool PwriteAll(const void* buffer, size_t size, off64_t offset) = 0;
+};
+
+class POSIXFileAccessor: public AbstractFileAccessor {
+  int fileDescriptor;
+public:
+  POSIXFileAccessor(int i = -1) : fileDescriptor{ i } {}
+  bool isValid() override { return fileDescriptor != -1; }
+  void invalidate() override { fileDescriptor = -1; }
+  void setFD(int i) override { fileDescriptor = i; }
+  int getAsFD() override { return fileDescriptor; }
+
+  int Open(const std::string &fname, int flags, mode_t createMask) override;
+  int FsyncAndClose() override;
+  int Close() override;
+  bool PreadAll(void* buffer, size_t size, off64_t offset) override;
+  bool PwriteAll(const void* buffer, size_t size, off64_t offset) override;
+};
 
 class FileReservationImpl : public khMTRefCounter {
   bool isWriter;
-  int fd;
+  std::unique_ptr<AbstractFileAccessor> aFA;
  public:
-  int Fd(void) const { return fd; }
-  FileReservationImpl(void) : isWriter(false), fd(-1) { }
+  AbstractFileAccessor *AFA(void) const { return aFA.get(); }
+  FileReservationImpl(void) : isWriter(false) { }
```

Comment:
It would be a good idea to create an empty POSIXFileAccessor here that is marked invalid and doesn't point to a file. That way you are guaranteed that you can dereference `aFA` without causing a memory error, even before you open the file.

Reply:
I'm not sure I'm following. This is meant to return the abstractfileaccessor member of filereservationimpl, which should already be instantiated as one of the derived classes.

Reply:
aFA doesn't get assigned until `FileReservationImpl::UnlockAndOpen_` is called, so the pointer returned by `aFA.get()` would be null, I believe. I think he's saying that you may want to set it to an initial value in the constructor. 
But to do that, it would have to be knowledgeable of `POSIXFileAccessor`, not just `AbstractFileAccessor`. Just brainstorming some options:
- Go ahead and create an empty `POSIXFileAccessor`, and just don't worry about it for now.
- Have a new type of accessor called `InvalidFileAccessor` that gets returned by a call to `getAccessor` if you pass in an empty file path